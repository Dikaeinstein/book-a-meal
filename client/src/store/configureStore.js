import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import throttle from 'lodash.throttle';
import { loadState, saveState } from '../helpers/localStorage';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers';

/**
 * Return redux middleware
 */
const getMiddleware = () => {
  const logger = createLogger();

  const middleware =
    process.env.NODE_ENV !== 'production'
    && process.env.NODE_ENV !== 'test'
      ? applyMiddleware(thunk, logger)
      : applyMiddleware(thunk);

  return middleware;
};

/**
 * Configure the redux store
 *
 * @returns {object} Redux store
 */
const configureStore = () => {
  const persistedState = loadState();

  const state = persistedState
    ? { ...initialState, ...persistedState }
    : initialState;

  const store = createStore(
    rootReducer,
    state,
    compose(
      getMiddleware(),
      window.devToolsExtension ? window.devToolsExtension() : empty => empty,
    ),
  );

  // Throttle the saveState function to be called once every second
  store.subscribe(throttle(() => {
    saveState({ orders: store.getState().orders });
  }, 1000));

  return store;
};

export default configureStore;
