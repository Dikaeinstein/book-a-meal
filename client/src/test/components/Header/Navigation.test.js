import React from 'react';
import { shallow } from 'enzyme';
import sweetAlert from 'sweetalert';
import { NavLink } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import ConnectedNavigation, { Navigation } from '../../../components/Header/Navigation';

jest.mock('sweetalert');
const signout = jest.fn();
sweetAlert.mockResolvedValue(Promise.resolve(true));

const urls = [
  {
    id: 1,
    name: 'Sign in',
    link: 'signin',
  },
  {
    id: 2,
    name: 'Sign up',
    link: 'signup',
  },
];

const customerUrls = [
  {
    id: 1,
    name: 'Menu',
    link: 'user-menu',
  },
  {
    id: 2,
    name: 'Orders',
    link: 'user-order-history',
  },
  {
    id: 3,
    name: 'Customer',
    link: 'user-details',
  },
  {
    id: 4,
    name: 'Logout',
    link: 'signout',
  },
];

const adminUrls = [
  {
    id: 1,
    name: 'Dashboard',
    link: 'dashboard',
  },
  {
    id: 2,
    name: 'Meals',
    link: 'meals',
  },
  {
    id: 3,
    name: 'Menus',
    link: 'menus',
  },
  {
    id: 4,
    name: 'Caterer',
    link: 'user-details',
  },
  {
    id: 5,
    name: 'Logout',
    link: 'signout',
  },
];

describe('<Navigation />', () => {
  beforeEach(() => {
    // values stored in tests will also be
    // available in other tests unless you run
    localStorage.clear();
  });
  it('should render 2 "NavLink"', () => {
    const userName = 'Dika';
    const wrapper = shallow(<Navigation
      urls={urls}
      userName={userName}
      signout={signout}
    />);
    expect(wrapper.find(NavLink).length).toEqual(urls.length);
  });
  it('should render NavLink with correct props', () => {
    const initialState = { urls: adminUrls, user: { data: { name: 'Dika Test' } } };
    const mockStore = configureStore();
    const store = mockStore(initialState);
    // Shallow render the container passing in the mock store
    const wrapper = shallow(<ConnectedNavigation
      store={store}
    />);
    wrapper.dive().find('li').last().simulate('click');
    expect(wrapper.props().urls).toEqual(adminUrls);
    expect(wrapper.props().userName).toEqual('Dika Test');
    expect(wrapper.props().signout).toBeDefined();
  });
  it('should render 3 "NavLink"', () => {
    const userName = 'Dika';
    const wrapper = shallow(<Navigation
      urls={adminUrls}
      userName={userName}
      signout={signout}
    />);
    expect(wrapper.find(NavLink).length).toEqual(3);
  });
  it('should render 2 "NavLink"', () => {
    const userName = 'Dika';
    const wrapper = shallow(<Navigation
      urls={customerUrls}
      userName={userName}
      signout={signout}
    />);
    expect(wrapper.find(NavLink).length).toEqual(2);
  });
  it('should <Navigation /> correctly', () => {
    const userName = 'Dika';
    const tree = shallow(<Navigation
      urls={urls}
      userName={userName}
      signout={signout}
    />);
    expect(tree).toMatchSnapshot();
  });
});
