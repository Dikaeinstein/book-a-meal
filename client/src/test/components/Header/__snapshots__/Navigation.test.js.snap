// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should <Navigation /> correctly 1`] = `
<nav
  className="col-3-4"
>
  <ul
    className="nav-container text-center"
  >
    <li
      key="1"
    >
      <NavLink
        activeClassName="active"
        ariaCurrent="true"
        to="/signin"
      >
        Sign in
      </NavLink>
    </li>
    <li
      key="2"
    >
      <NavLink
        activeClassName="active"
        ariaCurrent="true"
        to="/signup"
      >
        Sign up
      </NavLink>
    </li>
  </ul>
</nav>
`;
